typescript/sdk/src/ism/metadata/builder.ts

```diff
@@ -0,0 +1,140 @@
+import { TransactionReceipt } from '@ethersproject/providers';
+
+import {
+  WithAddress,
+  assert,
+  eqAddress,
+  rootLogger,
+} from '@hyperlane-xyz/utils';
+
+import { deepFind } from '../../../../utils/dist/objects.js';
+import { HyperlaneCore } from '../../core/HyperlaneCore.js';
+import { DispatchedMessage } from '../../core/types.js';
+import { DerivedHookConfigWithAddress } from '../../hook/read.js';
+import { HookType, MerkleTreeHookConfig } from '../../hook/types.js';
+import { DerivedIsmConfigWithAddress } from '../read.js';
+import { IsmType } from '../types.js';
+
+import {
+  AggregationIsmMetadata,
+  AggregationIsmMetadataBuilder,
+} from './aggregation.js';
+import { MultisigMetadata, MultisigMetadataBuilder } from './multisig.js';
+
+type NullMetadata = {
+  type:
+    | IsmType.PAUSABLE
+    | IsmType.TEST_ISM
+    | IsmType.OP_STACK
+    | IsmType.TRUSTED_RELAYER;
+};
+
+export type StructuredMetadata =
+  | AggregationIsmMetadata
+  | MultisigMetadata
+  | NullMetadata;
+
+export interface MetadataBuilder<
+  I extends DerivedIsmConfigWithAddress,
+  H extends DerivedHookConfigWithAddress,
+> {
+  build(
+    message: DispatchedMessage,
+    context: {
+      dispatchTx: TransactionReceipt;
+      hook: H;
+      ism: I;
+    },
+  ): Promise<string>;
+}
+
+export class BaseMetadataBuilder
+  implements
+    MetadataBuilder<DerivedIsmConfigWithAddress, DerivedHookConfigWithAddress>
+{
+  private multisigMetadataBuilder: MultisigMetadataBuilder;
+  private aggregationIsmMetadataBuilder: AggregationIsmMetadataBuilder;
+  protected logger = rootLogger.child({ module: 'BaseMetadataBuilder' });
+
+  constructor(protected readonly core: HyperlaneCore) {
+    this.multisigMetadataBuilder = new MultisigMetadataBuilder(core);
+    this.aggregationIsmMetadataBuilder = new AggregationIsmMetadataBuilder(
+      this,
+    );
+  }
+
+  // assumes that all post dispatch hooks are included in dispatchTx logs
+  async build(
+    message: DispatchedMessage,
+    context: {
+      dispatchTx: TransactionReceipt;
+      hook: DerivedHookConfigWithAddress;
+      ism: DerivedIsmConfigWithAddress;
+    },
+    maxDepth = 10,
+  ): Promise<string> {
```

Comment:
This is maybe premature, but would love to have Result type in here which makes surfacing errors to consumers a lot easier (like no quorum for example)